pkg/helm/chart.go

```diff
@@ -0,0 +1,267 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package helm
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/repo"
+
+	"github.com/kubernetes-sigs/minibroker/pkg/log"
+	"github.com/kubernetes-sigs/minibroker/pkg/nameutil"
+)
+
+// ChartInstaller is the interface that wraps the Install method.
+type ChartInstaller interface {
+	Install(
+		chartDef *repo.ChartVersion,
+		namespace string,
+		values map[string]interface{},
+	) (*release.Release, error)
+}
+
+// ChartUninstaller is the interface that wraps the Uninstall method.
+type ChartUninstaller interface {
+	Uninstall(releaseName, namespace string) error
+}
+
+// ChartInstallUninstaller wraps the ChartInstaller and ChartUninstaller interfaces.
+type ChartInstallUninstaller interface {
```

Comment:
No, we need these interfaces for mocking. Interfaces are not only for polymorphism.

Reply:
Yes, but I… didn't see any place where we mock, for example, `Uninstall()`? My grep finds:
```
pkg/helm/chart.go:43:// ChartUninstaller is the interface that wraps the Uninstall method.
pkg/helm/chart.go:45:   Uninstall(releaseName, namespace string) error
pkg/helm/chart.go:140:// Uninstall uninstalls a release from a namespace.
pkg/helm/chart.go:141:func (cc *ChartClient) Uninstall(releaseName, namespace string) error {
pkg/helm/chart.go:212:  actionNewUninstall func(*action.Configuration) *action.Uninstall
pkg/helm/chart.go:228:  actionNewUninstall func(*action.Configuration) *action.Uninstall,
pkg/helm/chart_test.go:256:             Describe("Uninstall", func() {
pkg/helm/chart_test.go:266:                             err := client.Uninstall(releaseName, namespace)
pkg/helm/chart_test.go:284:                             err := client.Uninstall(releaseName, namespace)
pkg/helm/chart_test.go:302:                             err := client.Uninstall(releaseName, namespace)
pkg/helm/chart_test.go:446:                             expectedUninstaller := &action.Uninstall{}
pkg/helm/chart_test.go:451:                             actionNewUninstall := func(arg0 *action.Configuration) *action.Uninstall {
pkg/minibroker/minibroker.go:717:       if err := c.helm.ChartUninstaller().Uninstall(releaseName, namespace); err != nil {
```
None of which is telling a mock to expect that call… `pkg/helm/chart_test.go:451` is `ChartUninstallRunner`, not `ChartInstallUninstaller`. `pkg/helm/chart_test.go:266`/`284`/`302` all mock `ChartHelmClientProvider.ProvideUninstaller()` instead of `ChartUninstaller.Uninstall()`.

I'm probably just missing something obvious :D

Reply:
No, it's not obvious. I didn't refactor the Minibroker package, where this is used. Nor did I create tests for the functions that use these interfaces. It will come soon.

Reply:
Then I'd argue that we want to put this in when it gets used, not now; currently it just makes things harder to review.

Reply:
I'm fine keeping or removing... fixed with 7bbf769.